tools/pal_inference.py

```diff
@@ -0,0 +1,471 @@
+    parser.add_argument(
+        '--max_length',
+        default=2048,
+        type=int,
+        help='Maximum input token length for the natural language description (default: 2048).'
+    )
+    parser.add_argument(
+        '--top_p',
+        default=0.8,
+        type=float,
+        help='Probability threshold to choose sample tokens during generation (default: 0.8).'
+    )
+    parser.add_argument(
+        '--temperature',
+        default=1,
+        type=float,
+        help='Temperature of token sampling during generation (default: 1).')
+    parser.add_argument(
+        '--time_out',
+        default=100,
+        type=float,
+        help='Maximum time allowed for executing generated code (default: 100).'
+    )
+    parser.add_argument(
+        '--verbose',
+        '-v',
+        action='store_true',
+        help='Print code error information when executing generated code (optional).'
+    )
+    parser.add_argument(
+        '--append',
+        '-a',
+        action='store_true',
+        help='Append output to the history results (optional).')
+    args = parser.parse_args()
+    return args
+
+
+@dataclass
+class GenerationConfig:
+    max_length: Optional[int] = None
+    top_p: Optional[float] = None
+    temperature: Optional[float] = None
+    do_sample: Optional[bool] = True
+    repetition_penalty: Optional[float] = 1.0
+
+
+@torch.inference_mode()
+def generate_interactive(
+    model: AutoModel,
+    tokenizer: AutoTokenizer,
+    prompt: str,
+    generation_config: Optional[GenerationConfig] = None,
+    logits_processor: Optional[LogitsProcessorList] = None,
+    stopping_criteria: Optional[StoppingCriteriaList] = None,
+    prefix_allowed_tokens_fn: Optional[Callable[[int, torch.Tensor],
+                                                List[int]]] = None,
+    additional_eos_token_id: Optional[int] = None,
+    **kwargs,
+):
+    inputs = tokenizer([prompt], padding=True, return_tensors='pt')
+    input_length = len(inputs['input_ids'][0])
+    for k, v in inputs.items():
+        inputs[k] = v.cuda()
+    input_ids = inputs['input_ids']
+    batch_size, input_ids_seq_length = input_ids.shape[0], input_ids.shape[
+        -1]  # noqa: F841
+    if generation_config is None:
+        generation_config = model.generation_config
+    generation_config = copy.deepcopy(generation_config)
+    model_kwargs = generation_config.update(**kwargs)
+    bos_token_id, eos_token_id = generation_config.bos_token_id, generation_config.eos_token_id  # noqa: F841
+    if isinstance(eos_token_id, int):
+        eos_token_id = [eos_token_id]
+    if additional_eos_token_id is not None:
+        eos_token_id.append(additional_eos_token_id)
+    has_default_max_length = kwargs.get(
+        'max_length') is None and generation_config.max_length is not None
+    if has_default_max_length and generation_config.max_new_tokens is None:
+        warnings.warn(
```

Comment:
still feel a little bit strange, can `logger.warn()` work here? if so, I think it can be unified